woocommerce/woocommerce-ios
# Product image upload: WP media library UI integration

UI integration part for #1875 

Sorry for the 800ish diff PR 🙇🏻‍♀️this PR integrates the UI for adding product images from WordPress Media Library: ~189 diffs are test cases on different media types, ~70 diffs on project/xib, and the largest file (with ~189 diffs) `WordPressMediaLibraryPickerDataSource.swift` has a bunch of empty implementation just to conform to `WPMediaGroup` and `WPMediaCollectionDataSource`.

## Changes

- Converted `Media` from `struct` to `NSObject` subclass in order to conform to `WPMediaAsset` 😞 (since `WPMediaPicker` is mostly in Objective-C)
  - Please check the next section for potential ways to improve how we fetch the image for a media item
  - In Yosemite:
    - Created `MediaType` enum that can be mapped to `WPMediaType` + unit tests
    - Added a `Media` extension `toProductImage` to convert a `Media` to `ProductImage` + unit test
- Created `WordPressMediaLibraryMediaGroup` that represents the collection of media items from the WP media library
- Created `WordPressMediaLibraryPickerDataSource` that implements `WPMediaCollectionDataSource` that:
  - specifies 1 media group (`WordPressMediaLibraryMediaGroup`)
  - retrieves the first page of WP media data on `loadData` that is called when loaded for the first time or pulled to refresh
  - fetches the next page of WP media data on the last cell of `media(at index: Int) -> WPMediaAsset` facilitated by `SyncingCoordinator` - potential improvement in the next section!
- Created `WordPressMediaLibraryImagePickerViewController` + xib that contains a fullscreen child view controller of type `WPNavigationMediaPickerViewController`
- In `MediaPickingCoordinator`, displayed the WP media picker UI when the user taps on the action sheet option. Updated `ProductImagesViewController` for `MediaPickingCoordinator` changes
- In `ProductImageActionHandler`, added `addSiteMediaLibraryImagesToProduct` to add product images from WP media library + unit test

## Future refactoring ideas

I'd like to improve a few things related to the current implementation of `WPMediaPicker` library, would love to hear your thoughts @shiki @pmusolino!
- Moved the image data provider (where it fetches the image data given for a media item) from [`WPMediaAsset` protocol](https://github.com/wordpress-mobile/MediaPicker-iOS/blob/develop/Pod/Classes/WPMediaCollectionDataSource.h#L95-L104) to somewhere else like `WPMediaCollectionDataSource`. This way, we could DI any dependencies (like `ImageService`) to the data source, instead of having to use the app singleton `ServiceLocator.imageService` to fetch the image in the `Media+WPMediaAsset` extension.
- Added paging support to `WPMediaPicker` by adding a new function like `mediaPickerController:willDisplayMediaForItemAtIndexPath:` to the `WPMediaPickerViewControllerDelegate`(https://github.com/wordpress-mobile/MediaPicker-iOS/blob/develop/Pod/Classes/WPMediaPickerViewController.h#L8-L19) protocol, and calling this function in [media view controller's `collectionView:willDisplayCell:forItemAtIndexPath:`](https://github.com/wordpress-mobile/MediaPicker-iOS/blob/develop/Pod/Classes/WPMediaPickerViewController.m#L993-L998) like how we usually trigger data fetching for the next page.

## Testing

- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- Tap "WordPress Media Library" option --> the WP media picker screen should be shown
- Scroll down to see the next page of images, if the site has more than 25 media items --> the images should be in the same order as in wp-admin > Media
- Pick some images and tap "Add {number of selected images}" in the navigation bar --> the selected images should show on the edit product images screen
- Tap "Done" on the edit product images screen --> it should go back to the edit product main screen
- Tap "Update" on the edit product main screen and wait for the remote update --> the images should be shown on the product page on web

## Example screenshots

screencast:

![simulator](https://user-images.githubusercontent.com/1945542/77885505-b32f3680-7299-11ea-926e-6eee1ab3d8db.gif)

placeholder | media picker - none selected | media picker - 2 selected
-- | -- | --
![Simulator Screen Shot - iPhone 11 - 2020-03-30 at 14 28 35](https://user-images.githubusercontent.com/1945542/77882497-a576b280-7293-11ea-9eea-be252fedde91.png) | ![Simulator Screen Shot - iPhone 11 - 2020-03-30 at 14 28 39](https://user-images.githubusercontent.com/1945542/77882502-a90a3980-7293-11ea-8768-8ea0d1dae4f1.png) | ![Simulator Screen Shot - iPhone 11 - 2020-03-30 at 14 29 23](https://user-images.githubusercontent.com/1945542/77885543-c80bca00-7299-11ea-8208-feb0eedfae6e.png)
 
Update release notes:
